include/deal.II/non_matching/quadrature_generator.h

```diff
@@ -0,0 +1,231 @@
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_non_matching_quadrature_generator
+#define dealii_non_matching_quadrature_generator
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/point.h>
+
+#include <type_traits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  namespace internal
+  {
+    namespace QuadratureGeneratorImplementation
+    {
+      /**
+       * Returns the center $(0.5, ..., 0.5)$ of the unit hypercube:
+       * $[0,1]^{dim}$.
+       */
+      template <int dim>
+      Point<dim>
+      unit_hypercube_center();
+
+
+      /**
+       * Describes a dim-dimensional hypercube aligned with the coordinate
+       * axes. That is, a region
+       *
+       * @f[
+       * [x_0-L/2, x_0+L/2] \times ... \times [x_{dim-1}-L/2, x_{dim-1}+L/2]
+       * @f]
+       *
+       * where L is the side length and $(x_0, ..., x_{dim-1})$ are the
+       * coordinates of the center of the hypercube.
+       *
+       * The purpose of this class is to help us keep track of where we are on
+       * the reference cell when generating immersed quadrature rules. There
+       * are two main things we want to be able to do.
+       *
+       * 1. Split a hypercube into its $2^{dim}$ children.
+       * 2. Take a cross section orthogonal to a given direction and get this
+       * as a Hypercube<dim-1>.
+       *
+       * These are needed because the algorithm is allowed to recurse both
+       * over children and over dimensions.
+       *
+       * In 3D, the 2 coordinates of the cross section of Hypercube<3> can be
+       * ordered in 2 ways. That is, if we take the cross section orthogonal
+       * to the y direction we could either order a 3D-coordinate into a
+       * 2D-coordinate as $(x,z)$ or as $(z,x)$. This class uses the second
+       * convention, corresponding to the coordinates being ordered cyclicly
+       * $x \rightarrow y \rightarrow z \rightarrow x \rightarrow ... $
+       * To be precise, if we take a cross section:
+       *
+       * | Orthogonal to | Cross section coordinates ordered as |
+       * |:-------------:|:------------------------------------:|
+       * |      x        |               (y, z)                 |
+       * |      y        |               (z, x)                 |
+       * |      z        |               (x, y)                 |
+       *
+       * This is according to the convention set by the function
+       * <code>coordinate_to_one_dim_higher</code>.
+       */
+      template <int dim>
+      class Hypercube
```

Comment:
I should have looked at this earlier, but isn't it the case that this class is in essence the same as `BoundingBox`? Couldn't one have extended `BoundingBox` by whatever additional functionality you need here?

Reply:
> Couldn't one have extended `BoundingBox` by whatever additional functionality you need here?

Yes, probably.  The member functions would be well defined for `BoundingBox` aswell.

I might be interested in the space-time case where the hypercube is 4-dimensional. I guess that would work with a `BoundingBox` too, but I'm not sure if `BoundingBox::get_neighbor_type` makes sense.